lib/cron/storage-events.js

```diff
@@ -0,0 +1,224 @@
+StorageEventsCron.FINALITY_TIME = 10800000; // 3 hours
+
+StorageEventsCron.prototype.start = function(callback) {
+  log.info('starting the storage events cron');
+
+  this.storage = new Storage(
+    this._config.storage.mongoUrl,
+    this._config.storage.mongoOpts,
+    { logger: log }
+  );
+
+  this.job = new CronJob({
+    cronTime: StorageEventsCron.CRON_TIME,
+    onTick: this.run.bind(this),
+    start: false,
+    timeZone: 'UTC'
+  });
+
+  this.job.start();
+
+  setImmediate(callback);
+};
+
+StorageEventsCron.prototype._resolveCodes = function(event, user) {
+  const threshold = this._config.application.unknownReportThreshold;
+
+  let success = event.success;
+  let successModified = false;
+  let unknown = success ? false : true;
+  if (unknown) {
+    resolveCodes();
+  }
+
+  function resolveCodes() {
+    const failureCode = 1100;
+
+    const clientCode = event.clientReport ?
+          event.clientReport.exchangeResultCode : undefined;
+    const farmerCode = event.farmerReport ?
+          event.farmerReport.exchangeResultCode : undefined;
+
+    /* jshint eqeqeq: false */
+    if (farmerCode == failureCode && !clientCode) {
+      success = false;
+      unknown = false;
+      return;
+    }
+
+    if (farmerCode == failureCode && clientCode == failureCode) {
+      success = false;
+      unknown = false;
+      return;
+    }
+
+    if (!farmerCode && clientCode == failureCode) {
+      success = false;
+      unknown = false;
+      return;
+    }
+
+    if (user.exceedsUnknownReportsThreshold(threshold)) {
+      successModified = true;
+      success = true;
+      unknown = true;
+      return;
+    }
+  }
+
+  return {
+    success: success,
+    successModified: successModified,
+    unknown: unknown
+  };
+
+};
+
+StorageEventsCron.prototype._resolveEvent = function(event, callback) {
+
+  this.storage.models.User.findOne({_id: event.user}, (err, user) => {
+    if (err) {
+      return callback(err);
+    }
+
+    const {success, successModified, unknown} = this._resolveCodes(event, user);
+
+    if (successModified) {
+      // TODO also give reputation points to farmer for successful
+      // transfer for the storage event
+      event.success = success;
+      event.save((err) => {
+        if (err) {
+          return callback(err);
+        }
+        finalize();
+      });
+    } else {
+      finalize();
+    }
+
+    function finalize() {
+      user.updateUnknownReports(unknown, event.timestamp, (err) => {
+        if (err) {
+          return callback(err);
+        }
+
+        callback(null, event.timestamp);
+      });
+    }
+  });
+};
+
+StorageEventsCron.prototype._run = function(lastTimestamp, callback) {
+
+  const StorageEvent = this.storage.models.StorageEvent;
+  const finalityTime = new Date(Date.now() - StorageEventsCron.FINALITY_TIME);
+
+  const cursor = StorageEvent.find({
+    timestamp: {
+      $lt: finalityTime,
+      $gt: lastTimestamp
```

Comment:
Two or more StorageEvents can have the same timestamp. If the job gets canceld by `StorageEventsCron.MAX_RUN_TIME` the lastTimestamp could be set at the first StorageEvent and restarting the job with  `$gt: lastTimestamp` would skip all other StorageEvents with the same timestamp.

I don't think `$gte` is a better choice. In this case it would not skip StorageEvents but it would read them twice :/

Reply:
It should be safe to repeat processing storage events within a certain degree (many hours). The event should remain unmodified, and the update to the user should not apply based on the timestamp: https://github.com/Storj/service-storage-models/pull/136/files#diff-6cde9fa75ac789afe7ff15f670cfc049R462